pkg/cmd/roachtest/roachtestutil/mixedversion/planner.go

```diff
@@ -683,90 +845,121 @@ func (p *testPlanner) changeVersionSteps(
-	})
+	p.setFinalizing(service, true)
+	p.setStage(service, RunningUpgradeMigrationsStage)
 
-	steps := []testStep{runSystemMigrations}
+	allowAutoUpgrade := p.newSingleStep(allowUpgradeStep{
+		virtualClusterName: service.Descriptor.Name,
+	})
 
-	var mixedVersionStepsDuringTenantMigrations []testStep
-	if scheduleHooks {
-		mixedVersionSteps := p.hooks.MixedVersionSteps(p.currentContext, p.prng, p.isLocal)
+	var steps []testStep
 
-		switch p.deploymentMode {
-		case SystemOnlyDeployment:
-			steps = append(steps, mixedVersionSteps...)
+	// runStepsWithTenantMigrations adds a step to `SET` the tenant's
+	// cluster version to the version we are upgrading to. In addition,
+	// if `scheduleHooks` is true, we choose a (possibly empty) subset
+	// of user-hooks to run concurrently with these migrations.
+	runStepsWithTenantMigrations := func() {
+		upgradeSteps := []testStep{
+			p.newSingleStep(setClusterVersionStep{
+				v:                  toVersion,
+				virtualClusterName: p.tenantName(),
+			}),
+		}
 
-		case SharedProcessDeployment:
-			// If we are in a multi-tenant deployment, we run some
-			// mixed-version hooks while the system is finalizing, and the
-			// remaining hooks while the tenant is finalizing.
-			if len(mixedVersionSteps) > 0 {
-				idx := p.prng.Intn(len(mixedVersionSteps))
+		if scheduleHooks {
+			upgradeSteps = append(upgradeSteps, p.hooks.MixedVersionSteps(p.currentContext, p.prng)...)
+		}
 
-				mixedVersionStepsDuringSystemMigrations := mixedVersionSteps[:idx]
-				steps = append(steps, mixedVersionStepsDuringSystemMigrations...)
+		steps = append(steps, p.concurrently(genericLabel, upgradeSteps)...)
+	}
 
-				mixedVersionStepsDuringTenantMigrations = mixedVersionSteps[idx:]
-			}
+	// If the service being upgraded is the system service, we allow
+	// auto-upgrades since they are supported.
+	if service.IsSystem() {
+		steps = []testStep{allowAutoUpgrade}
+		if scheduleHooks {
+			steps = append(
+				steps,
+				p.concurrently(mixedVersionLabel, p.hooks.MixedVersionSteps(p.currentContext, p.prng))...,
+			)
 		}
+	} else {
+		// In separate-process deployments, we set the cluster setting
+		// explicitly, as auto-upgrades are not supported. We also run
+		// all steps selected to run during finalization at this stage.
+		runStepsWithTenantMigrations()
 	}
 
 	steps = append(steps, p.newSingleStep(
 		waitForStableClusterVersionStep{
-			nodes:              p.currentContext.System.Descriptor.Nodes,
+			nodes:              service.Descriptor.Nodes,
 			timeout:            p.options.upgradeTimeout,
 			desiredVersion:     versionToClusterVersion(toVersion),
-			virtualClusterName: install.SystemInterfaceName,
+			virtualClusterName: service.Descriptor.Name,
 		},
 	))
 
-	// At this point, we just upgraded the storage cluster; we might
-	// need to patch up the `system.tenant_settings` table in order for
-	// the tenant upgrade to work.
+	// If this is a system-only or separate-process deployment, we are
```

Comment:
Nit: I found the "we are done" part confusing in case of separate-process. The fact is that this function is invoked twice, i.e., first with `p.currentContext.System`, then with `p.currentContext.Tenant`. It might be worth clarifying.